src/Symfony/Component/HttpClient/CachingHttpClient.php

```diff
@@ -106,41 +404,319 @@ public function request(string $method, string $url, array $options = []): Respo
+     *
+     * This function will return true if the server response can be cached,
+     * false otherwise.
+     *
+     * @param array<string, string|string[]> $requestHeaders
+     * @param array<string, string|string[]> $responseHeaders
+     * @param array<string, string|true>     $cacheControl
+     */
+    private function isServerResponseCacheable(int $statusCode, array $requestHeaders, array $responseHeaders, array $cacheControl): bool
+    {
+        // no-store => skip caching
+        if (isset($cacheControl['no-store'])) {
+            return false;
+        }
+
+        if (
+            $this->sharedCache
+            && !isset($cacheControl['public']) && !isset($cacheControl['s-maxage']) && !isset($cacheControl['must-revalidate'])
+            && isset($requestHeaders['authorization'])
+        ) {
+            return false;
+        }
+
+        if ($this->sharedCache && isset($cacheControl['private'])) {
+            return false;
+        }
+
+        // Conditionals require an explicit expiration
+        if (\in_array($statusCode, self::CONDITIONALLY_CACHEABLE_STATUS_CODES, true)) {
+            return $this->hasExplicitExpiration($responseHeaders, $cacheControl);
+        }
+
+        return \in_array($statusCode, self::CACHEABLE_STATUS_CODES, true);
+    }
+
+    /**
+     * Checks if the response has an explicit expiration.
+     *
+     * This function will return true if the response has an explicit expiration
+     * time specified in the headers or in the Cache-Control directives,
+     * false otherwise.
+     *
+     * @param array<string, string|string[]> $headers
+     * @param array<string, string|true>     $cacheControl
+     */
+    private function hasExplicitExpiration(array $headers, array $cacheControl): bool
+    {
+        return isset($headers['expires'])
+            || ($this->sharedCache && isset($cacheControl['s-maxage']))
+            || isset($cacheControl['max-age']);
+    }
+
+    /**
+     * Creates a MockResponse object from cached data.
+     *
+     * This function constructs a MockResponse from the cached data, including
+     * the original request method, URL, and options, as well as the cached
+     * response headers and content. The constructed MockResponse is then
+     * returned.
+     *
+     * @param array{chunks_count: int, status_code: int, initial_age: int, headers: array<string, string|string[]>, stored_at: int} $cachedData
+     */
+    private function createResponseFromCache(string $key, array $cachedData, string $method, string $url, array $options, string $fullUrlTag): MockResponse
+    {
+        return MockResponse::fromRequest(
+            $method,
+            $url,
+            $options,
+            new MockResponse(
+                (function () use ($key, $cachedData, $fullUrlTag): \Generator {
+                    for ($i = 0; $i <= $cachedData['chunks_count']; ++$i) {
+                        $chunkItem = $this->cache->getItem("{$key}_chunk_{$i}");
```

Comment:
maybe make the closure static and "use" `$cache`?

Reply:
`$cache` is a property of the class so we can't use "use" it in a closure. Do you mean you want to introduce a temp variable (eg. `$cache = $this->cache`) for it that I can then use in the closure ? I frankly don't understand why but if that's what you meant, I'll do it.